### DISCUSSION TITLE
anyone have deployment problem when use CRA with web worker ?

### DISCUSSION THREAD

**Discussion by XXXMrG** [vercel/vercel#4467] [Help]
I am working with react with web worker. used GoogleChromeLabs comlink as web worker expose way. also used [worker-plugin](https://github.com/GoogleChromeLabs/worker-plugin) to load worker.js.
it always worked until today. I commit a new deployment.but the page broken to a blank.the console error log is react prod error 103: "Uncaught Error: Minified React error # 130"
I try to rollback to a history version. but when old version redeployed also broken.
then i try run `vercel dev` locally. the console error message is: 
`Failed to load module script: The server responded with a non-JavaScript MIME type of "text/html". Strict MIME type checking is enforced for module scripts per HTML spec.`
it seem something wrong with worker.js loaded.
but magically. my worker.js file loaded way have worked right for a long time.
I have no idea about this problem. I also do not know why it happened.
I really need someone's hand.🙏
there is my repo: https://github.com/XXXMrG/recursive-animation
deployment site: https://recursive-animation.now.sh/

**Comment by hvaoc** [comment]
This is how I managed to get it done using worker-loader and webpack. You are getting a 404 when trying to load your worker.js in the browser. That's why you are getting a HTML instead of Javascript. That's what the MIME type complain actually mean. 

You may encounter this - So it may give you the impression that it worked and broke now. Fact is it  is broken and you just didn't notice it on all cases. When you jump from one page to another on the browner in SPA way, worker may load properly but when you server render it will break.

If you are using Next.js / Webpack you assets get rearranged and will not be available where you are expecting it to be. 

If you manage to make it work with worker-plugin, let me know. Hope the sample below will help you. 

```
// Source - https://github.com/zeit/next-plugins/blob/master/packages/next-workers/index.js
// Note - I am not using next-worker just got the sample config for Webpack from there.

const withCSS = require('@zeit/next-css')

module.exports = withCSS({
    /* config options here */
    webpack: (config) => {
        config.module.rules.push(
            {
                test   : /\.worker\.ts$/,
                loader : 'worker-loader',
                options: {
                    name      : 'static/[hash].worker.js',
                    publicPath: '/_next/'
                }
            })

        // Overcome Webpack referencing `window` in chunks
        config.output.globalObject = `(typeof self !== 'undefined' ? self : this)`

        return config
    }
})
```

```
/// testPage.tsx

import JankWorker from '../workers/jank.worker'

async componentDidMount() {
let jankWorker = new JankWorker()
        // console.log(jankWorker)
        // @ts-ignore
        jankWorker.postMessage('A message from the page')
        // @ts-ignore
        jankWorker.addEventListener('message', event => {
            const { data } = event
            console.log(data)
        })
}
```

If you want a Shared Worker, you need to patch to make it into one since worker-loader doesn't support SharedWorker.

```
// testPage.tsx

async componentDidMount() {

        // Monkey Patch from https://github.com/webpack-contrib/worker-loader/issues/2
        let createSharedWorkerClass = function(workerFn) {
            let fn = workerFn.toString()
            let fnBody = fn.slice(fn.indexOf('{') + 1, fn.lastIndexOf('}'))
            let fnBodyShared = '__webpack_require__ = {p:path};' + fnBody.replace('Worker', 'SharedWorker')
            return new Function('path', fnBodyShared)
        }

        const MySharedWorker = createSharedWorkerClass(JankWorker)
        // @ts-ignore
        this.sharedWorker = new MySharedWorker()

        this.sharedWorker.port.start()
        this.sharedWorker.port.postMessage({ method: 'mul', values: [1, 20] })

        this.sharedWorker.port.onmessage = function(e) {
            console.log('mySharedWorker.onmessage', e.data)
            console.log('Message received from mySharedWorker')
        }
    }
```

**Worker file**

```
// some.worker.ts - Worker must end with worker.ts for webpack to compile it properly.

// @ts-ignore
let text = 'Hello I am Your Web Worker'

// @ts-ignore
const sentData = () => {
    // @ts-ignore
    postMessage({ text })
}

// eslint-disable-next-line no-restricted-globals
onmessage = ({ data }) => {
    console.log(data)
    sentData()
}
```

**tsconfig.json** (if you need it)

```
{
  "compilerOptions": {
    "target": "es2015",
    "lib": [
      "dom",
      "dom.iterable",
      "esnext"
    ],
    "allowJs": true,
    "skipLibCheck": true,
    "strict": false,
    "forceConsistentCasingInFileNames": true,
    "noEmit": true,
    "esModuleInterop": true,
    "module": "esnext",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "jsx": "preserve",
    "experimentalDecorators": true
  },
  "include": [
    "next-env.d.ts",
    "**/*.ts",
    "**/*.tsx"
  ],
  "exclude": [
    "node_modules",
    ".next",
    "next.config.js",
    "server/**/*.ts"
  ]
}
```
